### ISSUE TITLE
Mongodb atlas is not connecting on cloudrun. Throwing Could not connect to any servers in your MongoDB Atlas cluster. One common reason is that you're trying to access the database from an IP that isn't whitelisted. Make sure your current IP address is on your Atlas cluster's IP whitelist: https://docs.atlas.mongodb.com/security-whitelist/","reason";:{"type":"ReplicaSetNoPrimary","servers":{},"stale":false,"compatible":true,"heartbeatFrequencyMS":10000,"localThresholdMS":15,"setName":"atlas-xxxx"}}

### ISSUE THREAD

**Issue by Bipin-Chaudhary** [Automattic/mongoose#13769]
### Prerequisites

- [X] I have written a descriptive issue title

### Mongoose version

6.10.5

### Node.js version

14.16.1

### MongoDB version

4.4.23

### Operating system

None

### Issue

We have a service running on GCP cloudrun with nodejs and mongoose. I recently had to upgrade mongoose from 5.12.13 to 6.10.5. after deploying the new revision to cloudrun, it started throwing this error - Could not connect to any servers in your MongoDB Atlas cluster. One common reason is that you're trying to access the database from an IP that isn't whitelisted. Make sure your current IP address is on your Atlas cluster's IP whitelist: [https://docs.atlas.mongodb.com/security-whitelist/","reason"](https://docs.atlas.mongodb.com/security-whitelist/%22,%22reason%22);:{"type":"ReplicaSetNoPrimary","servers":{},"stale":false,"compatible":true,"heartbeatFrequencyMS":10000,"localThresholdMS":15,"setName":"atlas-xxx"}}

In atlas we have kept open access to IP's 0.0.0.0/0

When I downgrade to mongoose@5.12.13 it connects without errors. It only gives error when we use mongoose@6.x version. the error is throwing at mongoose.createConnection(uri). I also tried to use mongoose.connect instead but the error still appears. In the error object server:{} is empty. It doesn't show any servers.

I have tried running the same code locally and it connects to atlas successfully but it throws this error when we deploy it to gcp cloudrun.

we really need to migrate our mongoose@5x to 6x but we are blocked by this issue. I would appreciate your help.

thanks in advance.

**Comment by vkarpov15** [comment]
Where you using `useNewUrlParser: true` and `useUnifiedTopology: true` options in Mongoose 5?

**Comment by Bipin-Chaudhary** [comment]
> Where you using `useNewUrlParser: true` and `useUnifiedTopology: true` options in Mongoose 5?

Yes, in mongoose 6 I have removed this options as they no longer supported.

**Comment by Bipin-Chaudhary** [comment]
I figured out the problem. We had added a "bson@1.14" in the package.json resolution section. 
"Bson" is a dependency of mongoose. Because of adding the package into resolution, it was installing this specific version for all our direct dependency packages. Bson@1.14 was compatible with the Mongoose 5. so when we migrated mongoose to 6x then also it was installing the bson@1.14 version. It was incompatible with the mongoose 6 which was leading to the error "MongodbServerSelection" error. It was a generic error and it was pointing to the ip whitelist issue. after hours of debugging we caught the error. it was "bson_1.BSON.serialize is not a function". this error generates because of incorrect bson dependency. so I removed the bson package from resolutions and then it installed the correct bson package. It solved the problem.
